dhis-2/dhis-api/src/main/java/org/hisp/dhis/programrule/ProgramRuleEvaluationEnvironment.java

```diff
@@ -0,0 +1,70 @@
+package org.hisp.dhis.programrule;
+
+/*
+ * Copyright (c) 2004-2019, University of Oslo
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlRootElement;
+import com.google.common.collect.Sets;
+import org.hisp.dhis.common.DxfNamespaces;
+
+import java.util.Set;
+
+/**
+ * @author Enrico Colasante
+ */
+@JacksonXmlRootElement( localName = "programRuleEvaluationEnvironment", namespace = DxfNamespaces.DXF_2_0 )
+public enum ProgramRuleEvaluationEnvironment
+{
+    WEB( "web" ),
+    ANDROID( "android" );
+
+    private final String value;
+
+    ProgramRuleEvaluationEnvironment( String value )
+    {
+        this.value = value;
+    }
+
+    public static ProgramRuleEvaluationEnvironment fromValue( String value )
+    {
+        for ( ProgramRuleEvaluationEnvironment type : ProgramRuleEvaluationEnvironment.values() )
+        {
+            if ( type.value.equalsIgnoreCase( value ) )
+            {
+                return type;
+            }
+        }
+
+        return null;
+    }
+
+    public static Set<ProgramRuleEvaluationEnvironment> getDefault()
```

Comment:
As discussed, consider wether a enum or two booleans can be used instead for simpler hibernate modelling.
Booleans might be more extendable of we add server or other environments.

Reply:
we can avoid BOTH. The web rule engine is only executed if ProgramRuleAction has either SEND MESSAGE OR SCHEDULE MESSAGE type. If any external engine ( Android ) executed rule engine and that end up in SENDMESSAGE then we can expose API on server side just to get the notificationtemplateId and render it over to notification service.

Reply:
Just to clarify: "WEB" refers to web-app frontend.

We are going to use this field/fields in 2 different places:
- To filter out for the user the actions that cannot be performed by one rule (at this moment you are right, there is no difference between web and android, they can perform the same actions, maybe in the future that could be different)
- To evaluate only the rules for the interested environment (some rules could make sense/be useful on the web but not on android)

Please @Bekkalizer correct me if I said anything wrong

Reply:
@enricocolasante ,

just to clarify, this is something the user configures; It's not hardcoded into the system. So to specify and include some examples to your listitems:

* Users want certain rules only to run on android/web due to context of the rule (IE. a warning message tailored for android)
* Users want certain rules only to run on android/web because the format doesn't fit in both places (IE. a warning is too big to easily diplay on android, so only show on web)


A second though, unrelated to the use-case: Instead of booleans, you could look into sets again, but implementing it as jsonb in the db. That way we can avoid joining tables in db and the model will be slightly cleaner and imo easier to maintain in the future. Not sure if this is too involved or not, just wanted to put it out there.

Reply:
First part is 100% clear.

Second part, I originally did what you are saying, but I didn't know about the jsonb, and I didn't want to join tables.
I think in this case jsonb this is the best way to go.